### PR TITLE
Advanced infusion provider

### PR DESCRIPTION
**Pull request by MCTBL**
make a new infusion provider, it like combine the **intercepter** and **old provider**
- it can order precise amount of essentia that matrix need
- it can transfer essentia immediately

https://github.com/user-attachments/assets/9de5e6ff-45fd-41e7-beb1-0c520bd98bd7

for higher resolution: https://discord.com/channels/181078474394566657/603348502637969419/1279887063583428720

### CODE REVIEW

**Comment by MCTBL** [comment]
When put the new provider down, it will search the matrix in 9x9x9 region.[searchMatrix()](https://github.com/GTNewHorizons/ThaumicEnergistics/pull/72/files#diff-ecc398371bc4237c8425238091d0139b961d88f9fae039fd2b18589de583d7d4R121)
After binding, it works like [Infusion Intercepter](https://github.com/GTNewHorizons/thaumicinsurgence/blob/master/src/main/java/thaumicinsurgence/tileentity/TileEntityInfusionIntercepter.java) from [thaumicinsurgence](https://github.com/GTNewHorizons/thaumicinsurgence).[grabAllAspects()](https://github.com/GTNewHorizons/ThaumicEnergistics/pull/72/files#diff-ecc398371bc4237c8425238091d0139b961d88f9fae039fd2b18589de583d7d4R157)
And when it detect the lack of which essentia, it will order precise amount of that essentia automatically.

File: src/main/java/thaumicenergistics/common/tiles/TileAdvancedInfusionProvider.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    @TileEvent(TileEventType.TICK)
+    public void onTick() {
+        if (this.matrix == null && this.isActive) {
+            if (this.matrixX == null || this.matrixY == null || this.matrixZ == null) {
+                this.searchMatrix();
+            } else {
+                this.bindMatrix();
+            }
```
Comment (OneEyeMaker) [inline src/main/java/thaumicenergistics/common/tiles/TileAdvancedInfusionProvider.java:228]: Search for matrix in 9x9x9 area should not be performed every tick. Add a counter variable and increment it every tick. And search for matrix every 10th or 20th tick.
```java
if (++tickCounter % 20 == 0) {
	if (this.matrixX == null || this.matrixY == null || this.matrixZ == null) {
        this.searchMatrix();
    }
}
```
You're also don't need to rebind matrix every tick - it is bound once during search procedure.

**Review by OneEyeMaker — CHANGES_REQUESTED** [review]

File: src/main/java/thaumicenergistics/common/tiles/TileAdvancedInfusionProvider.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+public class TileAdvancedInfusionProvider extends TileProviderBase implements IAspectSource {
+
+    private static final String NBT_MATRIX_X = "MatrixX", NBT_MATRIX_Y = "MatrixY", NBT_MATRIX_Z = "MatrixZ",
+            NBT_HAS_ADD_STABILITY = "HasAddStability";
+
+    public TileInfusionMatrix matrix = null;
+
+    private Integer matrixX = null, matrixY = null, matrixZ = null;
+
+    public boolean hasAddStability = false;
```
[the remark this answers is not in the document]
Comment (MCTBL) [reply src/main/java/thaumicenergistics/common/tiles/TileAdvancedInfusionProvider.java]: just forget to remove this

File: src/main/java/thaumicenergistics/common/tiles/TileAdvancedInfusionProvider.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    public boolean bindMatrix() {
+        return this.bindMatrix(this.matrixX, this.matrixY, this.matrixZ);
+    }
+
+    public void unbindMatrix() {
+        this.matrix = null;
+        this.matrixX = null;
+        this.matrixY = null;
+        this.matrixZ = null;
```
Comment (OneEyeMaker) [inline src/main/java/thaumicenergistics/common/tiles/TileAdvancedInfusionProvider.java:111]: I missed this previously, but after unbinding matrix all necessary values should be synced and saved to NBT. So `markForUpdate` and `saveChanges` methods should be called here.

Comment (MCTBL) [reply src/main/java/thaumicenergistics/common/tiles/TileAdvancedInfusionProvider.java:111]: Not your fault, I add this after your review😂
I will fix this later

**Comment by MCTBL** [comment]
new research and texture

https://github.com/user-attachments/assets/5b863811-a2cd-40b9-8624-f1ebe989d874

**Comment by EnderProyects** [comment]
You could also use the interceptor in the recipe, since is kinda an upgrade of this one

File: src/main/resources/assets/thaumicenergistics/lang/en_US.lang [hunk 4]
```diff
@@ -140,6 +145,11 @@ tc.research_name.thaumicenergistics.TEINFPROV=Infusion Provider
...
 thaumicenergistics.research_page.TEINFPROV.1=By researching the process of essentia infusion, you have discovered a method for instantaneous conversion between digi-sentia and essentia that can be used for infusion.<BR>By placing the §lInfusion Provider§r near a thaumic device, the device can draw essentia directly from the network as if it were a warded jar.<BR>This could prove to be very useful for the infusion alter.
 
+# Research Advanced Infusion Provider
+tc.research_name.thaumicenergistics.TEADVINFPROV=Advanced Infusion Provider
+tc.research_text.thaumicenergistics.TEADVINFPROV=easiest Infusion.
+thaumicenergistics.research_page.TEADVINFPROV.1=As you continue to optimize your ae network auto crafting, you've been stumped countless times by infusion recipes.<BR>Now, you have a new idea, why not leave the task of prepareing essentia to your ae network?<BR>So you developed a new infusion provider that includes the functionality of the Interceptor, and automatically orders the missing essentia. Also it can bind infusion matrix in a 9*9*9 area. Right click allows him to search the area for matrix and rebind it.
```
[the remark this answers is not in the document]
Comment (OneEyeMaker) [reply src/main/resources/assets/thaumicenergistics/lang/en_US.lang:151]: AE is name of the mod. ME is name of the network. Please, fix this in Chinese localization too.

**Comment by MCTBL** [comment]
> You could also use the interceptor in the recipe, since is kinda an upgrade of this one

that recipe is for ThE itself, GTNH's price should not that cheap🤣

**Review by OneEyeMaker — APPROVED** [review]
Looks good to me.

**Comment by Alastors** [comment]
What’s the point of this if it’s both slower than the intercepter, and doesn’t reduce the lag (which is the entire point of the intercepter.)

this just seems to place a worse hat on an already existing hat

**Comment by EnderProyects** [comment]
> this just seems to place a hat on an already existing hat

One of the things this one does better than the intercepter is that it can be placed anywhere near the altar, not just under it, so automation using storage busses or conduicts can be more hidden and seen better

**Comment by Alastors** [comment]
> > this just seems to place a hat on an already existing hat
> 
> 
> 
> One of the things this one does better than the intercepter is that it can be placed anywhere near the altar, not just under it, so automation using storage busses or conduicts can be more hidden and seen better

Like I said to MTL, this is an explicit design choice to make the intercepter more part of the infusion altar's multi.

**Pull request merged**